### PR TITLE
Add mish fusion transformation

### PR DESCRIPTION
**Pull request by iimironov**
Description: Mish nGraph transformation and MO implementation

JIRA: 32954, 33715

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests: Not added, we use simple eltwise shape inference and value propagation functions
* [x]  Framework operation tests: 33719
* [x]  Transformation tests: No new transformation added
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py: No new supported models
* [x]  Model Optimizer IR Reader check: Done

Documentation:
* [x]  Supported frameworks operations list: Done, operation was supported via transformation, now it supported directly
* [x]  Supported **public** models list: No new supported models
* [x]  New operations specification: Done, PR https://github.com/openvinotoolkit/openvino/pull/1149
* [x]  Guide on how to convert the **public** model: No new supported models
* [x]  User guide update: Nothing to update

### CODE REVIEW

**Comment by ilyachur** [comment]
Please fix CI.

Also I guess you should add a call of your transformation pass and integrate it to the PassManager, add new tests and etc.

File: inference-engine/src/transformations/include/transformations/mish_fusion.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+class TRANSFORMATIONS_API MishFusion;
+
+}  // namespace pass
+}  // namespace ngraph
+
+class ngraph::pass::MishFusion: public ngraph::pass::GraphRewrite {
+public:
+    MishFusion() : GraphRewrite() {
+        mish_fusion();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/mish_fusion.hpp:25]: Please put all the code into constructor

File: inference-engine/src/transformations/include/transformations/mish_fusion.hpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace ngraph {
+namespace pass {
+
+class TRANSFORMATIONS_API MishFusion;
+
+}  // namespace pass
+}  // namespace ngraph
+
+class ngraph::pass::MishFusion: public ngraph::pass::GraphRewrite {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/mish_fusion.hpp:22]: inherit from ngraph::pass::MatcherPass

File: inference-engine/src/transformations/src/transformations/mish_fusion.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    auto exp = std::make_shared<ngraph::opset4::Exp>(input0);
+    auto input_const = op::Constant::create(element::f64, Shape{1}, {-1});
+    auto add = std::make_shared<ngraph::opset4::Add>(exp, input_const);
+    auto log = std::make_shared<ngraph::opset4::Log>(add);
+    auto tanh = std::make_shared<ngraph::opset4::Tanh>(log);
+    auto mul = std::make_shared<ngraph::opset4::Multiply>(input0, tanh);
+
+    ngraph::graph_rewrite_callback callback = [](pattern::Matcher& m) {
+
+        auto mul = std::dynamic_pointer_cast<ngraph::opset4::Multiply> (m.get_match_root());
```
Comment (lazarevevgeny) [inline inference-engine/src/transformations/src/transformations/mish_fusion.cpp:24]: @GlebKazantaev , do we really need all these checks that the operations are of the specific type? Should pattern matcher take care of that?

**Comment by GlebKazantaev** [comment]
@iimironov Please replace your code with this and check that it works.
```
    auto input = ngraph::pattern::any_input();
    auto exp = std::make_shared<ngraph::opset4::Exp>(input);
    auto add = std::make_shared<ngraph::opset4::Add>(exp, ngraph::pattern::wrap_type<ngraph::opset4::Constant>());
    auto log = std::make_shared<ngraph::opset4::Log>(add);
    auto tanh = std::make_shared<ngraph::opset4::Tanh>(log);
    auto mul = std::make_shared<ngraph::opset4::Multiply>(input, tanh);

    ngraph::graph_rewrite_callback callback = [=](ngraph::pattern::Matcher& m) {
        auto & pattern_to_output = m.get_pattern_value_map();
        auto exp_input = pattern_to_output.at(input);
        
        auto mish = std::make_shared<ngraph::opset4::Mish>(exp_input);
        
        mish->set_friendly_name(m.get_match_root()->get_friendly_name());
        ngraph::copy_runtime_info({pattern_to_output.at(mul).get_node_shared_ptr(),
                                   pattern_to_output.at(tanh).get_node_shared_ptr(),
                                   pattern_to_output.at(log).get_node_shared_ptr(),
                                   pattern_to_output.at(add).get_node_shared_ptr(),
                                   pattern_to_output.at(exp).get_node_shared_ptr()}, mish);
        ngraph::replace_node(m.get_match_root(), mish);
        return true;
    };

    auto m = std::make_shared<ngraph::pattern::Matcher>(mul, "MishFusion");
    register_matcher(m, callback);
```

File: inference-engine/src/transformations/src/transformations/mish_fusion.cpp (hunk 3 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/mish_fusion.cpp:24]: See my answer below.

File: inference-engine/src/transformations/include/transformations/mish_fusion.hpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
+// Copyright (C) 2018-2020 Intel Corporation
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/mish_fusion.hpp:1]: Add detailed description for this transformation. You can find an example here: https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/transformations/include/transformations/optimize_strided_slice.hpp#L39

File: inference-engine/tests/functional/inference_engine/transformations/mish_fusion_test.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        auto input_const = ngraph::opset4::Constant::create(ngraph::element::f64, ngraph::Shape{1}, {-1});
+        auto add = std::make_shared<ngraph::opset4::Add>(exp, input_const);
+        auto log = std::make_shared<ngraph::opset4::Log>(add);
+        auto tanh = std::make_shared<ngraph::opset4::Tanh>(log);
+        auto mul = std::make_shared<ngraph::opset4::Multiply>(input0, tanh);
+
+        f = std::make_shared<ngraph::Function>(ngraph::NodeVector{mul}, ngraph::ParameterVector{input0});
+
+        ngraph::pass::Manager manager;
+        //manager.register_pass<ngraph::pass::VisualizeTree>("/home/imironov/projects/dpd_vcp_dl/svg_debug/before.svg");
```
Comment (lazarevevgeny) [inline inference-engine/tests/functional/inference_engine/transformations/mish_fusion_test.cpp:37]: Please, remove commented lines

File: inference-engine/src/transformations/src/transformations/mish_fusion.cpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+#include <memory>
+#include <vector>
+
+#include <ngraph/opsets/opset4.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+
+ngraph::pass::MishFusion::MishFusion() {
+    auto input = ngraph::pattern::any_input();
+        auto exp = std::make_shared<ngraph::opset4::Exp>(input);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/mish_fusion.cpp:16]: Please fix this indent to four spaces

File: inference-engine/src/transformations/src/transformations/mish_fusion.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        auto add = std::make_shared<ngraph::opset4::Add>(exp, ngraph::pattern::wrap_type<ngraph::opset4::Constant>());
+        auto log = std::make_shared<ngraph::opset4::Log>(add);
+        auto tanh = std::make_shared<ngraph::opset4::Tanh>(log);
+        auto mul = std::make_shared<ngraph::opset4::Multiply>(input, tanh);
+
+        ngraph::graph_rewrite_callback callback = [=](ngraph::pattern::Matcher& m) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/mish_fusion.cpp:22]: replace with `matcher_pass_callback`

File: inference-engine/src/transformations/src/transformations/mish_fusion.cpp [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+#include "transformations/mish_fusion.hpp"
+
+#include <memory>
+#include <vector>
+
+#include <ngraph/opsets/opset4.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+
+ngraph::pass::MishFusion::MishFusion() {
```
Comment (lazarevevgeny) [inline inference-engine/src/transformations/src/transformations/mish_fusion.cpp:14]: This transformation should be included into some list of transformation. Probably to `inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp`.

@GlebKazantaev, what do you think?

File: model-optimizer/extensions/ops/activation_ops.py [hunk 9]
```diff
@@ -225,3 +225,17 @@ def __init__(self, graph: Graph, attrs: dict):
...
+
+    def __init__(self, graph: Graph, attrs: dict):
+        mandatory_props = {
+            'op': self.op,
+            'type': None,
+            'in_ports_count': 1,
+            'out_ports_count': 1,
+            'infer': None
+        }
+        super().__init__(graph, mandatory_props, attrs)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/activation_ops.py:241]: Please, add empty end of line.

**Pull request merged**